### PR TITLE
Asynchronous support for TLS v1.3 TLSX ECC/DH key generation and key agreement

### PR DESCRIPTION
**Pull request by dgarske**
* Implemented async support in TLS v1.3 extension processing for ECC and DH key generation and agreement.
* Added async support to `SendTls13ClientHello`, `DoTls13ServerHello` and `DoTls13ClientHello`.
* Cleanup of the example client/server use key share code.

ZD 12065

### CODE REVIEW

File: src/tls.c [hunk 1]
```diff
@@ -7007,86 +7008,94 @@ static int TLSX_KeyShare_GenEccKey(WOLFSSL *ssl, KeyShareEntry* kse)
         case WOLFSSL_ECC_SECP521R1:
             curveId = ECC_SECP521R1;
             keySize = 66;
-            dataSize = keySize * 2 + 1;
             break;
         #endif /* !NO_ECC_SECP */
-    #endif
-    #if defined(HAVE_X448) && ECC_MIN_KEY_SZ <= 448
-        case WOLFSSL_ECC_X448:
```
Comment (SparkiDev) [inline src/tls.c:7015]: Why was this removed?

File: src/tls.c [hunk 2]
```diff
@@ -8205,7 +8218,7 @@ static int TLSX_KeyShare_SetSupported(WOLFSSL* ssl)
  * ssl  The SSL/TLS object.
  * returns 0 on success and other values indicate failure.
  */
-int TLSX_KeyShare_Establish(WOLFSSL *ssl)
+int TLSX_KeyShare_Establish(WOLFSSL *ssl, int* doHelloRetry)
```
Comment (SparkiDev) [inline src/tls.c:8221]: Add comment above for doHelloRetry.

File: src/tls13.c [hunk 3]
```diff
@@ -2835,19 +2848,55 @@ int SendTls13ClientHello(WOLFSSL* ssl)
...
+    {
+        /* Reset state */
+        ssl->options.asyncState = TLS_ASYNC_BEGIN;
+        XMEMSET(args, 0, sizeof(Sch13Args));
+    }
+
+    switch (ssl->options.asyncState) {
+    case TLS_ASYNC_BEGIN:
+    {
+    args->idx = RECORD_HEADER_SZ + HANDSHAKE_HEADER_SZ;
```
Comment (SparkiDev) [inline src/tls13.c:2869]: 4 spaces in.

File: src/tls.c [hunk 4]
```diff
@@ -6715,95 +6708,108 @@ static int TLSX_KeyShare_GenDhKey(WOLFSSL *ssl, KeyShareEntry* kse)
...
+                        (byte*)kse->privKey, &kse->keyLen, /* private */
+                        kse->pubKey, &kse->pubKeyLen /* public */
+                    );
+                }
+            }
+            else
+        #endif
+            {
+                /* Generate a new key pair */
+                ret = DhGenKeyPair(ssl, dhKey,
```
Comment (SparkiDev) [inline src/tls.c:6769]: This looks like the re-entrant code.
Shouldn't this be isolated from the code above?

File: src/tls.c [hunk 5]
```diff
@@ -7007,86 +7008,94 @@ static int TLSX_KeyShare_GenEccKey(WOLFSSL *ssl, KeyShareEntry* kse)
...
+                ret = wc_EccPrivateKeyDecode(keyDer->buffer, &idx, eccKey, keyDer->length);
+            }
+            else
+        #endif
+            {
+                /* set curve info for EccMakeKey "peer" info */
+                ret = wc_ecc_set_curve(eccKey, kse->keyLen, curveId);
+                if (ret == 0) {
+                    /* Generate ephemeral ECC key */
+                    ret = EccMakeKey(ssl, eccKey, eccKey);
```
Comment (SparkiDev) [inline src/tls.c:7054]: Re-entrant code needs to be isolated?

File: src/tls.c [hunk 6]
```diff
@@ -7572,55 +7551,73 @@ static int TLSX_KeyShare_ProcessEcc(WOLFSSL* ssl, KeyShareEntry* keyShareEntry)
...
     }
-#endif
 
-    do {
-    #if defined(WOLFSSL_ASYNC_CRYPT)
-        ret = wc_AsyncWait(ret, &keyShareKey->asyncDev, WC_ASYNC_FLAG_CALL_AGAIN);
+    if (ret == 0 && eccKey == NULL)
+        ret = BAD_FUNC_ARG;
+    if (ret == 0) {
+        ret = EccSharedSecret(ssl, eccKey, ssl->peerEccKey,
```
Comment (SparkiDev) [inline src/tls.c:7597]: Isolate?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/tls.c (hunk 1 above)
Comment (dgarske) [reply src/tls.c:7015]: It is not used for the x448 case. Dead code

File: src/tls13.c (hunk 3 above)
Comment (dgarske) [reply src/tls13.c:2869]: I did not add spacing to keep the line count low and so the diff is minimal.

File: src/tls.c (hunk 4 above)
Comment (dgarske) [reply src/tls.c:6769]: This part only gets called once. It is not re entrant.

File: src/tls.c (hunk 5 above)
Comment (dgarske) [reply src/tls.c:7054]: Same. Not re entrant code. Called and it's the poll and higher level function that handles the async response code.

File: src/tls.c (hunk 4 above)
Comment (SparkiDev) [reply src/tls.c:6769]: Ah. The comment you added helped me make sense of this. Thanks.

File: src/tls.c (hunk 6 above)
Comment (SparkiDev) [reply src/tls.c:7617]: Nope

File: src/tls.c [hunk 7]
```diff
@@ -6715,95 +6708,113 @@ static int TLSX_KeyShare_GenDhKey(WOLFSSL *ssl, KeyShareEntry* kse)
...
+        /* Cleanup on error, otherwise data owned by key share entry */
+        if (kse->privKey != NULL) {
+            XFREE(kse->privKey, ssl->heap, DYNAMIC_TYPE_PRIVATE_KEY);
+            kse->privKey = NULL;
+        }
+        if (kse->pubKey != NULL) {
+            XFREE(kse->pubKey, ssl->heap, DYNAMIC_TYPE_PUBLIC_KEY);
+            kse->pubKey = NULL;
+        }
+        if (dhKey != NULL)
```
Comment (SparkiDev) [inline src/tls.c:6812]: I would like dhKey freed anyway.
DhKey is large and for small memory usage I don't want it hanging around any longer than it has to.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/tls.c (hunk 7 above)
Comment (dgarske) [reply src/tls.c:6812]: @SparkiDev , Very good feedback to reduce memory use! I've made that fix and will push soon once done testing.

File: src/internal.c [hunk 8]
```diff
@@ -27786,17 +27809,22 @@ static int DoSessionTicket(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
         if (ssl->suites->suites[i]   == peerSuites->suites[j] &&
             ssl->suites->suites[i+1] == peerSuites->suites[j+1] ) {
 
-            if (VerifyServerSuite(ssl, i)) {
-                int result;
+            int ret = VerifyServerSuite(ssl, i);
+            #ifdef WOLFSSL_ASYNC_CRYPT
+            if (ret == WC_PENDING_E)
+                return ret;
+            else
```
Comment (SparkiDev) [inline src/internal.c:27816]: No need for 'else'.

File: src/tls.c [hunk 9]
```diff
@@ -7273,91 +7307,89 @@ static int TLSX_KeyShare_ProcessDh(WOLFSSL* ssl, KeyShareEntry* keyShareEntry)
...
-    if (ret == WC_PENDING_E) {
-        ret = wc_AsyncWait(ret, &dhKey->asyncDev, WC_ASYNC_FLAG_NONE);
-    }
-#endif
     /* RFC 8446 Section 7.4.1:
-     *     ... left-padded with zeros up to the size of the prime. ...
-     */
-    if (params->p_len > ssl->arrays->preMasterSz) {
-        word32 diff = params->p_len - ssl->arrays->preMasterSz;
+    *     ... left-padded with zeros up to the size of the prime. ...
```
Comment (SparkiDev) [inline src/tls.c:7360]: Put space back in before '*'.

File: src/tls.c [hunk 10]
```diff
@@ -8179,13 +8231,18 @@ static int TLSX_KeyShare_SetSupported(WOLFSSL* ssl)
     /* Delete the old key share data list. */
     extension = TLSX_Find(ssl->extensions, TLSX_KEY_SHARE);
     if (extension != NULL) {
-        TLSX_KeyShare_FreeAll((KeyShareEntry*)extension->data, ssl->heap);
+        KeyShareEntry* kse = (KeyShareEntry*)extension->data;
+    #ifdef WOLFSSL_ASYNC_CRYPT
+        if (kse && kse->lastRet == WC_PENDING_E)
+            return TLSX_KeyShare_GenKey(ssl, kse);
```
Comment (SparkiDev) [inline src/tls.c:8237]: Assuming this is completing the operation before freeing.
Can the code continue if the key generation finishes?

File: src/tls.c [hunk 11]
```diff
@@ -8309,6 +8387,15 @@ int TLSX_KeyShare_DeriveSecret(WOLFSSL *ssl)
     TLSX*          extension;
     KeyShareEntry* list = NULL;
 
+#ifdef WOLFSSL_ASYNC_CRYPT
+    ret = wolfSSL_AsyncPop(ssl, NULL);
+    if (ret != WC_NOT_PENDING_E) {
+        /* Check for error */
+        if (ret < 0)
```
Comment (SparkiDev) [inline src/tls.c:8394]: move up to statement above.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: examples/client/client.c [hunk 12]
```diff
@@ -2996,43 +3015,7 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
...
-        if (onlyKeyShare == 0 || onlyKeyShare == 1) {
-        #ifdef HAVE_FFDHE_2048
-            if (wolfSSL_UseKeyShare(ssl, WOLFSSL_FFDHE_2048)
-                                                           != WOLFSSL_SUCCESS) {
-                err_sys("unable to use DH 2048-bit parameters");
-            }
-        #endif
-        }
-    #endif
+        SetKeyShare(ssl, onlyKeyShare, useX25519, useX448, 0);
```
[the remark this answers is not in the document]
Comment (dgarske) [reply examples/client/client.c:3018]: It was not called with the original code, so keeping existing behavior. Should the client be calling `wolfSSL_set_groups` is all cases?

File: src/tls.c (hunk 10 above)
Comment (dgarske) [reply src/tls.c:8237]: No, this is completing the previous key generation done in `TLSX_KeyShare_Use`. I've fixed this to instead fall-through and call `TLSX_KeyShare_Use` again. Thanks, much cleaner.

File: examples/client/client.c [hunk 13]
```diff
@@ -3373,7 +3356,7 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
         if (bio != NULL) {
             if (wolfSSL_SESSION_print(bio, wolfSSL_get_session(ssl)) !=
                     WOLFSSL_SUCCESS) {
-                wolfSSL_BIO_printf(bio, "ERROR: Unable to print out session\n");
+                wolfSSL_BIO_printf(bio, "BIO error printing session\n");
```
Comment (dgarske) [inline examples/client/client.c:3359]: Note: Changed to avoid using "ERROR:" text, which incorrectly causes async reports in the Jenkins job.

File: src/internal.c [hunk 14]
```diff
@@ -5129,9 +5131,16 @@ int DhAgree(WOLFSSL* ssl, DhKey* dhKey,
 #endif
     {
 #if !defined(HAVE_FIPS) && !defined(HAVE_SELFTEST)
-        ret = wc_DhCheckPubValue(ssl->buffers.serverDH_P.buffer,
-                    ssl->buffers.serverDH_P.length, otherPub, otherPubSz);
+        if (dhKey != NULL && (prime == NULL || primeSz == 0)) {
+            ret = wc_DhCheckPubKey(dhKey, otherPub, otherPubSz);
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/internal.c]: Had to put this change back to as it was due to missing prime values in the TLS v1.3 case. The `wc_DhCheckPubKey` does not do exponentiation so leaving as-is.

**Comment by dgarske** [comment]
@SparkiDev all feedback taken in and ready for next round of review. Thank you

File: examples/client/client.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (SparkiDev) [reply examples/client/client.c:3018]: wolfSSL_set_groups() will set an extension that indicates the supported curves on the client.
No need to set it unless strictly requiring the groups.
Benchmarking should strictly require the groups it is testing, so both calls should call it.
The other call shouldn't as the user may want to send key shares that aren't supported by the server and have the server respond with a HelloRetry.

So change the call in ClientBenchmarkThroughput() to pass 1.
